### PR TITLE
Remove special case unpartitioned-to-time mapping in automaterialization

### PR DESCRIPTION
**Pull request by psarka**
## Summary & Motivation

Removes special casing that ignores missing time partitions that descend from unpartitioned asset. This has been replaced by throttling mechanism driven by materializations per minute. Fixes https://github.com/dagster-io/dagster/issues/16033.

## How I Tested These Changes

I reran the repro provided in the issue and observed the correct behavior.

### CODE REVIEW

**Comment by psarka** [comment]
To clarify what has changed, this is how things looked before this change and after this change (ref #16033):

no deps eager 1:
before: ✔️ last partition
after: ✔️ last partition

no deps eager 100:
before: ✔️ all partitions in parallel
after: ✔️ all partitions in parallel

self dep eager 1:
before: ✔️ all partitions sequentially
after: ✔️ all partitions sequentially

self dep eager 100:
before: ✔️ all partitions sequentially
after: ✔️ all partitions sequentially

upstream dep eager 1:
before: ✔️ last partition
after: ✔️ last partition

upstream dep eager 100:
before: 🛑 last partition
after: ✔️ all partitions in parallel                        !! Fixed !!

self dep upstream dep eager 1:
before: 🛑 no materializations
after: ✔️ all partitions sequentially                    !! Fixed !!

self dep upstream dep eager 100:
before: 🛑 no materializations
after: ✔️ all partitions sequentially                    !! Fixed !!

**Comment by psarka** [comment]
Also, to my understanding, the changed behavior was not covered by unit tests. From the first glance it seems tricky to test this, but I can give a shot if you think that it is worthwhile and if you can give some hints on how to approach this.

**Comment by OwenKephart** [comment]
Hi @psarka, thanks for taking the initiative and looking into this, and sorry for the delay in response on the original issue. This extra information makes it a lot clearer what's going on here, and I think informs the correct path forward here.

If we were to remove this check, the end result is that any update to the unpartitioned asset (even after the backfill is completed) would result in a run for ALL downstream partitioned assets being kicked off. In this case, I don't believe that this is your desired behavior. At the very least, for the majority of users, the expectation is the current behavior, where that unpartitioned update will only impact the most recent partition of the downstream.

If I'm understanding correctly, your desired behavior only really applies when the partitions of the downstream are _missing_, which aligns with the following existing issue: https://github.com/dagster-io/dagster/issues/15617.

In essence, we currently only evaluate missing conditions from scratch for root assets in the graph. However, if we fixed that behavior, then your downstream partitioned asset would attempt to materialize all missing partitions (as long as the unpartitioned parent existed). Does that match your understanding?

**Comment by psarka** [comment]
Thank you for engaging with this PR @OwenKephart! (btw, I'm based in Europe, so also replying with a delay, was sleeping already. If you reply early in your morning I should be able to react faster)

I ran a test and I could not replicate this statement:

> If we were to remove this check, the end result is that any update to the unpartitioned asset [...] would result in a run for ALL downstream partitioned assets being kicked off.

Instead, this is what I'm getting (After rerunning upstream unpartitioned asset (x in my repro) manually):

upstream dep eager 1:
before: ✔️ last partition
after: ✔️ last partition

upstream dep eager 100:
before: 🛑 last partition
after: 🛑 last partition                >> I interpret this as broken, and it is broken the same way now

self dep upstream dep eager 1:
before: 🛑 no materializations
after: ✔️ last partition                 >> Fixed

self dep upstream dep eager 100:
before: 🛑 no materializations
after: 🛑 last partition                 >> I interpret this as broken, and it is slightly less broken than now

Do you mean something else by "update"? I interpreted it as "rematerialize".

As for the rest of the comment and relation to https://github.com/dagster-io/dagster/issues/15617, I would like to first understand what is the current behavior, and what is the intended behavior.

**Comment by OwenKephart** [comment]
Hi @psarka -- just got back from some PTO, thanks for your patience here, apologies for timezone difficulties.

> Do you mean something else by "update"? I interpreted it as "rematerialize".

I mean "rematerialize" here (I tend to use "update" as a general term to also cover changes to observable source assets, but that's not relevant for this issue)

> upstream dep eager 100:
before: 🛑 last partition
after: 🛑 last partition𝅷𝅷𝅷𝅷 .................. >> I interpret this as broken, and it is broken the same way now

I see how this can be interpreted as broken in this case, although this is intended behavior (and I now realize that I did not originally understand your original desired behavior). This goes back to "for the majority of users, the expectation is the current behavior, where that unpartitioned update will only impact the most recent partition of the downstream".

This is essentially a modeling issue, where Dagster does not currently have a way to distinguish between two different scenarios where you have unpartitioned -> time-partitioned assets:

1. This unpartitioned asset represents a rarely-changing asset that all time partitions depend on. An update to this asset should result in all partitions being updated.
2. This unpartitioned asset represents an append-only asset that different time partitions consume. An update to this asset will not impact previously-materialized partitions.

Option (2) is the more common scenario, and so the existing implementations cater to this in the absence of some framework-level information to pull on.

> self dep upstream dep eager 1:
before: 🛑 no materializations
after: ✔️ last partition𝅷𝅷𝅷𝅷 .................. >> Fixed

> self dep upstream dep eager 100:
before: 🛑 no materializations
after: 🛑 last partition𝅷𝅷𝅷𝅷 .................. >> I interpret this as broken, and it is slightly less broken than now

These both lead back to that same modeling issue above, where it's unclear at the framework level if historical partitions actually have any reason to be recomputed in response to an upstream unpartitioned asset being materialized.

I definitely do think this is a hole in the system, but in a world where we can only pick one behavior, I think the behavior as it stand is the better choice (purely in terms of the relative frequency of the two cases).

**Comment by psarka** [comment]
Sorry, I'm not following you :(

Does my proposed fix change some behavior from desirable to broken? No matter how you interpret 🛑 last partition𝅷𝅷𝅷𝅷 in the upstream dep eager 100 "rematerialize" case, it is unaffected by this PR.

**Comment by OwenKephart** [comment]
Got it -- I see what you mean, and this behavior does seem to work as you're describing, although I'll have to dig into this a bit more to understand why that is and if there are any side effects here, as you're correct that this is definitely contrary to my expectations. In general we definitely tend to be pretty cautious on any changes in this realm. Thanks for engaging on this, I'll get back to you on this tomorrow.

**Comment by OwenKephart** [comment]
I was able to look into this and figure out what's happening here. Essentially, the `materialize_on_parent_updated` rule operates in two phases:

1. Get the set of partitions whose parents have been updated since the last tick.
2. Check which specific parents have updated (so that we can render this information in the UI)

This PR causes step (1) to go from containing just the last partition of time-partitioned asset to containing all partitions.

In general, it's not expected that anything should be filtered out by step (2) because you shouldn't have been able to get to that step without a parent updating, but in this case things actually do end up getting filtered out. This is because the logic for checking which parents were updated mirrors the "map_old_time_partitions" which was removed in this PR, but in reverse (as it's looking for parents rather than children).

Then I looked into:

> self dep upstream dep eager 1:
before: 🛑 no materializations
after: ✔️ last partition𝅷𝅷𝅷𝅷 .................. >> Fixed

And I noticed that this seems to be the current behavior without modifications. In addition to trying out your repro on the current version of dagster and observing the expected results, I wrote up a unit test (`exotic_partition_mapping_scenarios.py`):

```python
# a self-dependent asset downstream of an unpartitioned asset
unpartitioned_self_dependency = [
    asset_def("asset1"),
    asset_def(
        "asset2",
        partitions_def=DailyPartitionsDefinition(start_date="2020-01-01"),
        deps={
            "asset1": None,
            "asset2": TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
        },
    ),
]

...

    "unpartitioned_self_dependency": AssetReconciliationScenario(
        assets=unpartitioned_self_dependency,
        cursor_from=AssetReconciliationScenario(
            assets=unpartitioned_self_dependency,
            unevaluated_runs=[
                # fully backfill the self-dependent asset
                run(["asset1"]),
                run(["asset2"], partition_key="2020-01-01"),
                run(["asset2"], partition_key="2020-01-02"),
                run(["asset2"], partition_key="2020-01-03"),
            ],
            # no runs necessary
            expected_run_requests=[],
        ),
        # now re-materialize the unpartitioned asset
        unevaluated_runs=[
            run(["asset1"]),
        ],
        # this results in the latest partition of the self-dependent asset being updated
        expected_run_requests=[run_request(asset_keys=["asset2"], partition_key="2020-01-03")],
        current_time=create_pendulum_time(year=2020, month=1, day=4, hour=4),
    ),

```

This unit test passes before and after your changes.

So to summarize, my view is that this PR does not actually modify behavior (and so it does not degrade or improve correctness, if we're on the same page that `LAST` is the correct default option). The reason that it does not modify behavior is somewhat of a coincidence, and the new version of this code would have at least some negative performance impact (causing us to need to iterate through partitions that we know will not pass the filter).

If that tracks with you, I'll close this PR and we can follow up on a new issue re: potential changes to enable behavior that would make sense for you (if there are remaining concerns).

**Comment by psarka** [comment]
It is hard to make progress, so I'm fine with closing this PR. I opened a more focused issue as per your suggestion. Some comments:

1. You write that you managed to get self_dep_upstream_dep asset to automaterialize. It does not work for me. I'm simply running dagster dev, enabling auto and nothing happens. I tried to materialize manually and rematerialize upstream, still nothing. Issue #16890 focuses just on this case, so maybe you will be able to share how you do it. 

2. I do not understand why you do so much ad-hoc filtering in a Rule called `MaterializeOnParentUpdated`. Reading the code I took that the guiding principle for automaterialization is "many small rules, neatly composed". For me all the ad-hoc cases are red flags indicating that somewhere rigor got lost and code is turning into a patch-a-mole. My instinct is to jump on this, unravel the thread and clean things up.

3. My expectation from "many small rules, neatly composed" was that in this case filtering will be done by `DiscardOnMaxMaterializationsExceeded`. I thought that this is the rule that solves the ALL vs LAST question. If I want to rerun all of my partitions when upstream updates, I set max_materializations to ∞. If I want only last, I set max_materializations to 1. But I'm probably misunderstanding it.

4. I would not like my issue to be resolved by yet another ad-hoc `if: else:`. I will deploy a patched version of Dagster, so don't hesitate to just say that you won't fix it.

5. If you see a prospect of pulling this thread and cleaning things up, then please keep in mind that I'm eager to help. I admit that this PR was a bit too brash (but it illustrated the point well), I can instead do more testing, try to synthesize a spec, find more bugs, etc. Let me know :)

**Comment by OwenKephart** [comment]
@psarka You raise some great points here, but just going point by point:

1. Here's my repro:
```python
from dagster import (
    AssetIn,
    AssetKey,
    AutoMaterializePolicy,
    DailyPartitionsDefinition,
    Nothing,
    TimeWindowPartitionMapping,
    asset,
)


@asset(auto_materialize_policy=AutoMaterializePolicy.eager(1000))
def x() -> str:
    return "hi"


@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-10-01"),
    auto_materialize_policy=AutoMaterializePolicy.eager(1),
    ins={
        "x": AssetIn(
            key=AssetKey("x"),
        ),
        "self_dep_upstream_dep": AssetIn(
            key=AssetKey("self_dep_upstream_dep"),
            partition_mapping=TimeWindowPartitionMapping(start_offset=-1, end_offset=-1),
            dagster_type=Nothing,
        ),
    },
)
def self_dep_upstream_dep(x: str) -> None:
    pass
```
- (starting with brand-new instance, although this shouldn't matter)
- materialize unpartitioned asset
- note that first materialization of the unpartitioned asset will not kick off anything for the self-dep asset, as this will be seen as a signal to update the last partition of the self-dep asset, which can't materialize yet (because the previous day isn't there)
- backfill the self-dep asset
- re-materialize the unpartitioned asset
- this time, I see the last partition get updated:
Screenshot 2023-10-05 at 4 55 44 PM

Not sure exactly what's diverging between our repros here, I did also try with the gist you provided and observed a similar sequence of events, just tried to slim things down to the basics here.

2. Totally agree that the materialize_on_parent_updated rule is quite complicated, and this comes from two sources. One is this underlying friction in the API, where different parts of the codebase need to make different assumptions about what to do with unpartitioned -> partitioned mappings (as has been causing problems on your end). The other source is the need for this rule to be performant (as this is called very often), and so many tricks are used in order to minimize the amount of querying / processing that needs to be done. It definitely could use some streamlining and another pass to tidy things up though.
3. That's also a fair interpretation. The main issue would just be that `DISCARD`s are intended to indicate that something happened which the user should take some manual action to resolve (e.g. we have considered in the future letting users alert whenever discards happen, or to give ways for people to backfill any discarded partitions). This approach would muddy the mental model somewhat, by having people expect partitions to be discarded regularly.
4. Agree that an if/else would not be the way to go here, and the full solution would likely be along the lines of some PartitionMapping-type concept that can be applied to the mapping between unpartitioned and partitioned assets, but that's not something we're currently planning on implementing in the near future.
5. Thank you! We're definitely open to PRs in this area, and it's possible that there's something better than what I suggested in point 4 which would make sense to iterate on. Definitely open to ideas and future PRs.

Again, really appreciate your patience here, it's been a busy couple weeks on our side as we've just pushed out a major release (1.5)!
